### ISSUE TITLE
Issue building oneDNN 2.6 with MinGW-w64

### ISSUE THREAD

**Issue by brechtsanders** [oneapi-src/oneDNN#1332]
oneDNN 2.6 wouldn't built with MinGW-w64 due to assembler issues.
It was trying to use MASM (`ml`), which is not part of MinGW-w64 gcc/binutils.
I was able to build after making the following change:
```
patch -ulbf src/common/CMakeLists.txt << EOF
@@ -33,3 +33,3 @@
             set(CMAKE_C_FLAGS "\${CMAKE_C_FLAGS} -DITT_API_IPT_SUPPORT")
-            if(NOT WIN32)
+            if(NOT MSVC)
                 enable_language(ASM)
EOF
```

**Comment by vpirogov** [comment]
Thanks for the report, @brechtsanders!

**Comment by igorsafo** [comment]
Hi @brechtsanders , thanks for your report. Could you please verify the fix?
- rls-v2.6: #1352
- master: #1351 

Regards,
Igor

**Comment by brechtsanders** [comment]
Building from master gives:
´´´
[27/381] Building ASM_MASM object src/...mmon.dir/ittnotify/ittptmark64.asm.obj
FAILED: src/common/CMakeFiles/dnnl_common.dir/ittnotify/ittptmark64.asm.obj
ml -DDNNL_ENABLE_CPU_ISA_HINTS -DDNNL_ENABLE_ITT_TASKS -DDNNL_ENABLE_MAX_CPU_ISA -DDNNL_X64=1 -DNOMINMAX -D_WIN -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -IR:/winlibs64-11.2.0ucrt/oneDNN/include -IR:/winlibs64-11.2.0ucrt/oneDNN/build_static/include -IR:/winlibs64-11.2.0ucrt/oneDNN/src  -c -Fo src/common/CMakeFiles/dnnl_common.dir/ittnotify/ittptmark64.asm.obj R:/winlibs64-11.2.0ucrt/oneDNN/src/common/ittnotify/ittptmark64.asm
CreateProcess failed: The system cannot find the file specified.
ninja: fatal: ReadFile: The handle is invalid.
´´´
MinGW-w64 is Windows, but uses GCC, not MSVC, so the assembler should not be `ml`.

**Comment by igorsafo** [comment]
@brechtsanders Sorry, i am confused. What branch did you use to validate the fix?
In the PR I added a check for minGW to fix it: https://github.com/oneapi-src/oneDNN/pull/1351 
Please note that the fix **is not promoted yet**, so building from master will result in an issue.

**Comment by brechtsanders** [comment]
I'm on Windows and I don't know (yet) how to check out a specific PR. What is the command line for that?

Also, why does this PR force using pthreads on Windows when using MinGW? Native Windows threads should work well. Or is there a reason for forcing the use of pthreads?

**Comment by igorsafo** [comment]
In the PR #1352 there is a source branch which is igorsafo:rls-v2.6/fix-itt-mingw, which means that the branch is in a fork. To try that branch:
```bash
git clone https://github.com/igorsafo/oneDNN.git # checkout the fork
cd oneDNN
git checkout rls-v2.6/fix-itt-mingw # switch to the branch with the fix
# build oneDNN
# ...
```

The reason for the second commit which forces to use posix threading when mingw is used  is that I tried to reproduce the issue locally. During oneDNN build I found that minGW with native Windows threads doesn't support `std::thread` and `std::promise`, which are used in oneDNN. Actually I can remove the last comment to fix the issue you are seeing and not breaking your build. What minGW do you use?

**Comment by brechtsanders** [comment]
Thanks for the instructions.

Building with `-DDNNL_BUILD_TESTS:BOOL=ON` fails (as it did before)
```
tests/benchdnn/dnnl_common.cpp:774:10: fatal error: sys/sysinfo.h: No such file or directory
```
So I can't run the tests. I also built without python or doxyrest.

Both static and shared builds work perfectly.

I use MinGW-w64, more specifically my personal build from https://winlibs.com/

**Comment by igorsafo** [comment]
Hi @brechtsanders , the fixes landed into both rls-v2.6 and master branches.
